lib/spack/spack/spec.py

```diff
@@ -1642,7 +1642,12 @@ def to_node_dict(self, hash=ht.dag_hash):
                 d['patches'] = variant._patches_in_order_of_appearance
 
         if hash.package_hash:
-            d['package_hash'] = self.package.content_hash()
+            package_hash = self.package.content_hash()
+
+            # Full hashes are in bytes
+            if isinstance(package_hash, bytes):
```

Comment:
Is this intended to convert all hashes to `unicode` type in python 2.7?  It may not actually affect anything, but I'm always wary of `isinstance(..., bytes)` after some issues with 2/3 compat in flux.

Reply:
Is there a better way to do it? The full hash (unlike the others) is in bytes so we do need to convert it (somehow!)

Reply:
The trick is that `bytes` and `str` are the same type in python2.7, so this will convert all strings in python 2 to the `unicode` type but only `bytes` instances to `str` in python3.  As long as it's ok that all of them come out unicode I'd probably leave it like this, perhaps with a comment about it.  If we want everything to be the native text type `str` regardless of the python version, then the check should probably be something like `if (not isinstance(package_hash, six.text_type)) and isinstance(package_hash, six.binary_type)`.

Reply:
I'm happy to change that! @trws do you have familiarity with hooks and know how we could keep state, for something like spack monitor?  My questions here -> https://github.com/spack/spack/pull/21930#discussion_r597056059. It's what I'm working on now and need to figure out before updating the PR.